diagnosis/clusterhealthcheck.sh

```diff
@@ -0,0 +1,230 @@
+
+log_level -i "Checking for kubernetes containers running on nodes"
+MASTER_CONTAINERS="kube-scheduler kube-controller kube-apiserver kube-addon-manager"
+NODE_CONTAINERS="azure-ip-masq-agent kube-flannel kube-proxy"
+for IP in $HOST_LIST
+do
+    HOST_NAME=$(ssh -q -t $USER_NAME@$IP 'echo "$(hostname)"')
+    CONTAINER_LIST=$(ssh -q -t $USER_NAME@$IP 'LIST=$(docker container ls --format="{{.Names }}" --filter status="running"); echo $LIST')
+    
+    if [[ $HOST_NAME == *"master"* ]]; then
+        log_level -i "Checking for master containers"
+        for CONTAINER in $MASTER_CONTAINERS
+        do
+            if [[ $CONTAINER_LIST == *"$CONTAINER"* ]]; then
+                log_level -i "$CONTAINER running on [$HOST_NAME]. Test passed"
+            else
+                log_level -e "$CONTAINER not running on [$HOST_NAME]. Test failed"
+            fi
+        done
+        log_level -i "Checking for node containers"
+        for CONTAINER in $NODE_CONTAINERS
+        do
+            if [[ $CONTAINER_LIST == *"$CONTAINER"* ]]; then
+                log_level -i "$CONTAINER running on [$HOST_NAME]. Test passed"
+            else
+                log_level -e "$CONTAINER not running on [$HOST_NAME]. Test failed"
+            fi
+        done
+    else
+        log_level -i "Checking for node containers"
+        for CONTAINER in $NODE_CONTAINERS
+        do
+            if [[ $CONTAINER_LIST == *"$CONTAINER"* ]]; then
+                log_level -i "$CONTAINER running on [$HOST_NAME]. Test passed"
+            else
+                log_level -e "$CONTAINER not running on [$HOST_NAME]. Test failed"
+            fi
+        done
+    fi
+done
+
+# ##############################################################
+# Check if node to node communication works
+
+log_level -i "Checking nodes can communicate with other nodes"
+for IP in $HOST_LIST
+do
+    HOST_NAME=$(ssh -q -t $USER_NAME@$IP 'echo "$(hostname)"')
+    log_level -i "Checking connectivity from [$HOST_NAME] to cluster nodes"
+    for CURRENTIP in $HOST_LIST
+    do
+        if [[ $IP != $CURRENTIP ]]; then
+            PING_RESULT=$(ssh -q -t $USER_NAME@$IP "ping -c 1 -W 1 "$CURRENTIP" | grep '64 bytes'")
+            if [[ ! -z $PING_RESULT ]]; then
+                log_level -i "Connectivity from $HOST_NAME to $CURRENTIP successfull"
+            else
+                log_level -e "Connectivity from $HOST_NAME to $CURRENTIP failed"
+            fi
+        fi
+    done
+done
+
+# ##############################################################
+# Checking for outboud connectivity on nodes
+
+log_level -i "Checking if nodes have connection to the internet"
+for IP in $HOST_LIST
+do
+    HOST_NAME=$(ssh -q -t $USER_NAME@$IP 'echo "$(hostname)"')
+    
+    log_level -i "Checking connectivity from [$HOST_NAME] to internet"
+    NET_RESULT=$(ssh -q -t $USER_NAME@$IP 'wget -q --spider http://google.com; if [ $? -eq 0 ]; then echo "Online"; else echo "Offline"; fi')
```

Comment:
AKSe uses `netcat` to test outbound connectivity

`nc -vz docker.io 443`

Reply:
will look into this